File: blog-backend/handler/ip_blacklist.go
package handler

import (
	"blog-backend/db"
	"blog-backend/model"
	"blog-backend/util"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

type IPBlacklistHandler struct{}

func NewIPBlacklistHandler() *IPBlacklistHandler {
	return &IPBlacklistHandler{}
}

// List 获取IP黑名单列表
func (h *IPBlacklistHandler) List(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	var blacklist []model.IPBlacklist
	var total int64

	offset := (page - 1) * pageSize

	// 获取总数
	db.DB.Model(&model.IPBlacklist{}).Count(&total)

	// 分页查询
	err := db.DB.Order("created_at DESC").
		Limit(pageSize).
		Offset(offset).
		Find(&blacklist).Error

	if err != nil {
		util.Error(c, 500, err.Error())
		return
	}

	util.Success(c, gin.H{
		"list":      blacklist,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
	})
}

// Add 添加IP到黑名单
func (h *IPBlacklistHandler) Add(c *gin.Context) {
	var req struct {
		IP       string `json:"ip" binding:"required"`
		Reason   string `json:"reason"`
		Duration int    `json:"duration"` // 封禁时长（小时），0表示永久
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		util.BadRequest(c, "参数错误")
		return
	}

	// 检查IP格式
	if !util.IsValidIP(req.IP) {
		util.BadRequest(c, "IP地址格式不正确")
		return
	}

	var expireAt *time.Time
	if req.Duration > 0 {
		expire := time.Now().Add(time.Duration(req.Duration) * time.Hour)
		expireAt = &expire
	}

	blacklist := model.IPBlacklist{
		IP:       req.IP,
		Reason:   req.Reason,
		BanType:  2, // 手动封禁
		ExpireAt: expireAt,
	}

	// 检查是否已存在
	var existing model.IPBlacklist
	if err := db.DB.Where("ip = ?", req.IP).First(&existing).Error; err == nil {
		util.Error(c, 400, "该IP已在黑名单中")
		return
	}

	if err := db.DB.Create(&blacklist).Error; err != nil {
		util.Error(c, 500, err.Error())
		return
	}

	util.SuccessWithMessage(c, "添加成功", blacklist)
}

// Delete 从黑名单中删除IP
func (h *IPBlacklistHandler) Delete(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		util.BadRequest(c, "无效的ID")
		return
	}

	if err := db.DB.Delete(&model.IPBlacklist{}, id).Error; err != nil {
		util.Error(c, 500, err.Error())
		return
	}

	util.SuccessWithMessage(c, "删除成功", nil)
}

// Check 检查IP是否在黑名单中
func (h *IPBlacklistHandler) Check(c *gin.Context) {
	ip := c.Query("ip")
	if ip == "" {
		util.BadRequest(c, "请提供IP地址")
		return
	}

	var blacklist model.IPBlacklist
	err := db.DB.Where("ip = ?", ip).First(&blacklist).Error

	if err != nil {
		util.Success(c, gin.H{
			"banned": false,
		})
		return
	}

	// 检查是否过期
	if blacklist.ExpireAt != nil && blacklist.ExpireAt.Before(time.Now()) {
		db.DB.Delete(&blacklist)
		util.Success(c, gin.H{
			"banned": false,
		})
		return
	}

	util.Success(c, gin.H{
		"banned": true,
		"info":   blacklist,
	})
}

// CleanExpired 清理过期的黑名单记录
func (h *IPBlacklistHandler) CleanExpired(c *gin.Context) {
	result := db.DB.Where("expire_at IS NOT NULL AND expire_at < ?", time.Now()).
		Delete(&model.IPBlacklist{})

	if result.Error != nil {
		util.Error(c, 500, result.Error.Error())
		return
	}

	util.SuccessWithMessage(c, "清理成功", gin.H{
		"deleted_count": result.RowsAffected,
	})
}

File: blog-frontend/src/api/ipBlacklist.ts
/*
 * 项目名称：blog-frontend
 * 文件名称：ipBlacklist.ts
 * 创建时间：2026-02-01 19:52:47
 *
 * 系统用户：Administrator
 * 作　　者：無以菱
 * 联系邮箱：huangjing510@126.com
 * 功能描述：IP黑名单管理相关 API 接口定义，包括IP黑名单的增删改查、IP状态检查、过期记录清理等功能（管理员操作）。
 */

import { request } from '@/utils/request'
import type { PageData } from '@/types/common'

/**
 * IP黑名单接口
 */
export interface IPBlacklist {
  id: number                // 记录ID
  ip: string                // IP地址
  reason: string            // 封禁原因
  ban_type: number          // 封禁类型：1表示自动封禁，2表示手动封禁
  expire_at: string | null  // 过期时间，null表示永久封禁
  created_at: string        // 创建时间
  updated_at: string        // 更新时间
}

/**
 * IP检查结果接口
 */
export interface IPCheckResult {
  banned: boolean           // 是否被封禁
  info?: IPBlacklist        // 封禁信息（如果被封禁）
}

/**
 * 获取IP黑名单列表（管理员）
 * @param params 分页参数
 * @param params.page 页码（可选）
 * @param params.page_size 每页数量（可选）
 * @returns 返回分页的IP黑名单列表
 */
export function getIPBlacklist(params: { page?: number; page_size?: number }) {
  return request.get<PageData<IPBlacklist>>('/admin/ip-blacklist', { params })
}

/**
 * 添加IP到黑名单（管理员）
 * @param data IP黑名单数据
 * @param data.ip IP地址（必填）
 * @param data.reason 封禁原因（可选）
 * @param data.duration 封禁时长（小时），0表示永久（可选）
 * @returns 返回创建的IP黑名单记录
 */
export function addIPBlacklist(data: {
  ip: string
  reason?: string
  duration?: number // 封禁时长（小时），0表示永久
}) {
  return request.post<IPBlacklist>('/admin/ip-blacklist', data)
}

/**
 * 删除IP黑名单（管理员）
 * @param id 记录ID
 * @returns 返回删除结果
 */
export function deleteIPBlacklist(id: number) {
  return request.delete(`/admin/ip-blacklist/${id}`)
}

/**
 * 检查IP是否在黑名单中（管理员）
 * @param ip IP地址
 * @returns 返回IP检查结果
 */
export function checkIPStatus(ip: string) {
  return request.get<IPCheckResult>('/admin/ip-blacklist/check', { params: { ip } })
}

/**
 * 清理过期的黑名单记录（管理员）
 * @returns 返回清理结果，包含删除的记录数量
 */
export function cleanExpiredIPBlacklist() {
  return request.post<{ deleted_count: number }>('/admin/ip-blacklist/clean-expired')
}

